ocular/src/chain/client/query.rs

```diff
@@ -20,31 +20,43 @@ pub type SlashingQueryClient = slashing::query_client::QueryClient<Channel>;
 pub type StakingQueryClient = staking::query_client::QueryClient<Channel>;
 
 impl ChainClient {
-    pub fn check_for_grpc_address(&self) -> Result<(), GrpcError> {
-        if self.config.grpc_address.is_empty() {
-            return Err(GrpcError::MissingEndpoint(format!(
-                "no grpc address available for chain {}",
-                self.config.chain_id
-            )));
+    // Auth queries
+    pub async fn get_auth_query_client(&mut self) -> Result<AuthQueryClient, ChainClientError> {
+        let mut result: Result<AuthQueryClient, ChainClientError> =
+            Err(TxError::BroadcastError(String::from("Client connection never attempted.")).into());
+
+        // Get grpc address randomly each time; shuffles on failures
+        for _i in 0u8..self.connection_retry_attempts + 1 {
```

Comment:
Indeed we do, the problem is inside the loop we need to attempt a connection with the specific client (e.g. AuthQueryClient in this instance below), since they're all different types I dont think we can abstract the method unless they share some trait i missed.

Reply:
Ah. We can actually make a trait and them implement it on each of them and then use it as the method param.

Reply:
So, it turns out async trait methods aren't a thing yet in vanilla rust. There's a crate that allows for async traits, however there's issues passing futures around. So in order to use the chain client's get_random_grpc_endpoint() method and access the cache, we need to pass around a reference to the chain client in the trait implementation and make it look something like 

```
use async_trait::async_trait;

#[async_trait]
pub trait GrpcClient {
    async fn get_client(&self, connection_retry_attempts: u8, chain_client: &ChainClient) -> Result<Box<Self>, ChainClientError>;
}
```

Unfortunately this leads to some issues with the keystore somehow (I'm guessing some sort of thread safety mechanism) 
```
future cannot be sent between threads safely
the trait `std::marker::Sync` is not implemented for `(dyn keyring::KeyStore + 'static)`
required for the cast to the object type `dyn futures::Future<Output = std::result::Result<Box<cosmos_sdk_proto::cosmos::auth::v1beta1::query_client::QueryClient<tonic::transport::Channel>>, ChainClientError>> + std::marker::Send`rustc
query.rs(25, 63): captured value is not `Send` because `&` references cannot be sent unless their referent is `Sync`
```

All this to say, It maybe possible to work through, however I think the end result will be a bit messier going with the trait approach.